File: LRUCache.cpp
class LRUCache {
  int caps;
  unordered_map < int, pair < int, list < int > ::iterator > > * cache;
  list < int > * cache_list;
  public:

    LRUCache(int capacity) {
      caps = capacity;
      cache = new unordered_map < int, pair < int, list < int > ::iterator > > ();
      cache_list = new list < int > ();
    }

  int get(int key) {
    auto iter = cache -> find(key);
    if (iter == cache -> end())
      return -1;
    int val = iter -> second.first;
    cache_list -> erase(iter -> second.second);
    cache -> erase(iter);
    cache_list -> push_front(key);
    cache -> insert({
      key,
      {
        val,
        cache_list -> begin()
      }
    });

    if (cache_list -> size() > caps) {
      int k = cache_list -> back();
      cache -> erase(k);
      cache_list -> pop_back();
    }
    return val;
  }

  void put(int key, int value) {
    auto iter = cache -> find(key);
    if (iter != cache -> end()) {
      cache_list -> erase(iter -> second.second);
      cache -> erase(iter);
    }
    cache_list -> push_front(key);
    cache -> insert({
      key,
      {
        value,
        cache_list -> begin()
      }
    });

    if (cache_list -> size() > caps) {
      int k = cache_list -> back();
      cache -> erase(k);
      cache_list -> pop_back();
    }
  }
};

File: LRUCache.py
import collections
class LRUCache(object):

def __init__(self, capacity):
    """
    :type capacity: int
    """
    self.capacity = capacity
    self.lruCache = collections.OrderedDict()
    

def get(self, key):
    """
    :type key: int
    :rtype: int
    """
    if key in self.lruCache:
        # If we are getting the value, make it a recent entry to avoid eviction.
        popped = self.lruCache.pop(key)
        self.lruCache[key] = popped 
        return popped
    else:
        return -1
    

def put(self, key, value):
    """
    :type key: int
    :type value: int
    :rtype: None
    """
    if key in self.lruCache:
        self.lruCache.pop(key) 
        # key is already present, popped the old value, that would reduce cache as well for new entry.
    elif self.capacity <= len(self.lruCache): 
        #otherwise check size of cache first and if cache is full, pop LRU item Queue cache. 
        self.lruCache.popitem(last = False) 
    # Insert new key/value pair
    self.lruCache[key] = value

